external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala

```diff
@@ -73,12 +73,14 @@ class DirectKafkaInputDStream[
 
   protected val kc = new KafkaCluster(kafkaParams)
 
-  protected val maxMessagesPerPartition: Option[Long] = {
-    val ratePerSec = context.sparkContext.getConf.getInt(
+  private val ratePerSec: Int = context.sparkContext.getConf.getInt(
       "spark.streaming.kafka.maxRatePerPartition", 0)
-    if (ratePerSec > 0) {
+  protected def maxMessagesPerPartition: Option[Long] = {
+    val effectiveRatePerSec =
+      rateController.map(_.getLatestRate()).filter(_ > 0).fold(ratePerSec) {_ min _}
```

Comment:
Can't reproduce, but I think I know why : you're trying to compile with a scala version which doesn't match the POM, aren't you ? Probably even 2.11 ?